src/core/server/http/integration_tests/lifecycle_handlers.test.ts

```diff
@@ -0,0 +1,197 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { resolve } from 'path';
+import * as kbnTestServer from '../../../../test_utils/kbn_server';
+import { clusterClientMock } from './core_service.test.mocks';
+
+const pkgPath = resolve(__dirname, '../../../../../package.json');
+const actualVersion = require(pkgPath).version;
+
+const versionHeader = 'kbn-version';
+const xsrfHeader = 'kbn-xsrf';
+const nameHeader = 'kbn-name';
+
+const whitelistedTestPath = '/xsrf/test/route/whitelisted';
+
+const kibanaName = 'my-kibana-name';
+
+describe('core lifecycle handlers', () => {
+  let root: ReturnType<typeof kbnTestServer.createRoot>;
+
+  beforeEach(async () => {
+    root = kbnTestServer.createRoot({
+      server: {
+        name: kibanaName,
+        customResponseHeaders: {
+          'some-header': 'some-value',
+        },
+        xsrf: { disableProtection: false, whitelist: [whitelistedTestPath] },
+      },
+    });
+    await root.setup();
+    await root.start();
+  }, 30000);
+
+  afterEach(async () => {
+    clusterClientMock.mockClear();
+    await root.shutdown();
+  });
+
+  describe('versionCheck post-auth handler', () => {
+    const testRoute = '/version_check/test/route';
+
+    beforeEach(() => {
+      kbnTestServer.getKbnServer(root).server.route({
+        path: testRoute,
+        method: 'GET',
+        handler: () => 'ok',
+      });
+    });
+
+    it('accepts requests with the correct version passed in the version header', async () => {
+      await kbnTestServer.request
+        .get(root, testRoute)
+        .set(versionHeader, actualVersion)
+        .expect(200, 'ok');
+    });
+
+    it('accepts requests that do not include a version header', async () => {
+      await kbnTestServer.request.get(root, testRoute).expect(200, 'ok');
+    });
+
+    it('rejects requests with an incorrect version passed in the version header', async () => {
+      await kbnTestServer.request
+        .get(root, testRoute)
+        .set(versionHeader, 'invalid-version')
+        .expect(400, /"Browser client is out of date/);
+    });
+  });
+
+  describe('customHeaders pre-response handler', () => {
+    const testRoute = '/custom_headers/test/route';
+
+    beforeEach(() => {
+      kbnTestServer.getKbnServer(root).server.route({
+        path: testRoute,
+        method: 'GET',
+        handler: () => 'ok',
+      });
+    });
+
+    it('adds the kbn-name header', async () => {
+      const result = await kbnTestServer.request.get(root, testRoute).expect(200, 'ok');
+      const headers = result.header as Record<string, string>;
+      expect(headers).toEqual(
+        expect.objectContaining({
+          [nameHeader]: kibanaName,
+        })
+      );
+    });
+
+    it('adds the custom headers', async () => {
```

Comment:
could you add a test that custom headers are present in error responses as well?